File: Semestr 3/Paradygmaty_Programowania/Ocaml/cw_1.ml
(* zadanie 1 *)
let rec flatten list =
	if list = [] then []
	else List.hd list @ flatten (List.tl list)
	
(* zadanie 2 *)
let rec count (x, list) = 
	if list = [] then 0
	else if List.hd list = x then 1 + count (x, List.tl list)
	else count (x, List.tl list)
	
(* zadanie 3 *)
let rec replicate (a, x) = 
	if x < 0 then failwith "Niepoprawna liczba powtórzeń!"
	else if x = 0 then []
	else [a] @ replicate(a, x-1)
	
(* zadanie 4 *)	
let rec sqrList list =
	if list = [] then []
	else [List.hd list * List.hd list] @ sqrList(List.tl list)

(* zadanie 5 *)
let palindrome list =
	list = List.rev list
	
let rec palindrome2 list = 
	if List.length list < 2 then true
	else if not (List.hd list = List.hd (List.rev list)) then false
	else palindrome2 (List.tl(List.rev(List.tl list)))
	
(* zadanie 6 *)
let rec listLength list = 
	if list = [] then 0
	else 1 + listLength(List.tl list)
	
	

File: Semestr 3/Paradygmaty_Programowania/Scala/src/cw_1.scala
===SCALA===

// zadanie 1
def flatten[A](list:List[List[A]]):List[A] = 
  if (list == Nil) List()
  else list.head ++ flatten(list.tail)
 
// zadanie 2
def count[A](p:(A,List[A])):Int =
  if (p._2 == Nil) 0
  else if (p._2.head == p._1) 1 + count(p._1, p._2.tail)
  else count(p._1, p._2.tail)
  
// zadanie 3  
 def replicate[A](p:(A,Int)):List[A] = 
   if (p._2 < 0) throw new Exception("Nieprawidłowa liczba powtórzeń!")
   else if (p._2 == 0) List()
   else List(p._1) ++ replicate(p._1, p._2 - 1)

// zadanie 4
def sqrList(list:List[Int]):List[Int] = 
  if (list == Nil) List()
  else List(list.head * list.head) ++ sqrList(list.tail)
  
// zadanie 5
def palindrome[zg] (list:List[zg]) = list == list.reverse
  
// zadanie 6
def listLength[A] (list:List[A]):Int =
  if (list == Nil) 0
  else 1 + listLength(list.tail)

  
  
  